KEP: Coscheduling

k82cn opened the pull request:
Signed-off-by: Da K. Ma <klaus1982.cn@gmail.com>

/cc @vishh, @bsalamat, @mindprince, @jlewi 

/cc @kubernetes/sig-scheduling-feature-requests

etiennecoutaud commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,172 @@
+
+## References
+
+* [Gang scheduling in Kuberentes](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit#heading=h.ckn8nv2jj0xv)
```
typo Kuberentes /Kubernetes

bsalamat commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,174 @@
+    // this field from PriorityClassName.
+    // The higher the value, the higher the priority.
+    // +optional
+    Priority *int32
```
I think you should remove this field from the template. Priority is supposed to be resolved at the time of Pod creation. It  exists in PodSpec, but it doesn't make sense to be in the template. This field is not expected to be copied to PodSpec.

k82cn replied:
Reusing Priority for "Job" here; and the priority of Pod will only take effect within a "Job". 

There's the requirements on Job priority, IMO: https://github.com/kubernetes/kubernetes/issues/65746

bsalamat commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,174 @@
+    // fit on a node. Selector which must match a node's labels for the pod of
+    // 'Job' to be scheduled on that node.
+    // +optional
+    NodeSelector map[string]string
```
NodeSelector is in a way an earlier and inferior version of NodeAffinity. Why not use NodeAffinity instead of NodeSelector?

k82cn replied:
Agree, `NodeAffinity` is better; using NodeSelector because of simplification in prototype :)

bsalamat commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+  2. `Number of Running pods` >= `.spec.SchedSpec.minAvailable`: If there are still enough running pods for the `QueueJob`, the `QueueJobController` will only recreate failed pods, and wait for scheduler to bind the pending pods.
+
+* **QueueJob Update/Delete**: There's not rolling update for `QueueJob` ; if pod template was updated, the whole `QueueJob` is re-created. If `QueueJob` was deleted, all its pods and `SchedulingSpec` will be cleanup.
```
Do we plans to support running long-running services as a gang (QueueJob) as well? If we don't, we may not need to support rolling updates for batch jobs which are expected to run to completion.

k82cn replied:
> Do we plans to support running long-running services as a gang (QueueJob) as well?

Good question! I used to plan to support long-running services too; but for now, did not get requirement right now. So I'd like to make it batch work only firstly to make first release simple.

/cc @mindprince , fyi, some changes since we last sync up.

k82cn commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+* **Pod Failures/Deletion/Unschedulable**: When a pod was failed, deleted or unschedulable, `QueueJobController` manages `QueueJob`'s lifecycle according to its status:
+
+  1. `Number of Running pods` < `.spec.SchedSpec.minAvailable` : If the number of running pods less then `.spec.SchedSpec.minAvailable`, the `QueueJobController` will kill the whole QueueJob and recreate pods for rescheduling in batch.
```
Yes; it need to include `Succeeded` pods: `Succeeded + Runinning <= minAvailable`

bsalamat commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,174 @@
+![queuejob](images/gang-scheduling-queuejob.png)
+
+
+### Customized controller
```
We also need to cover admission in this document. I think we ideally want to admit either the whole gang or none of the members of a gang. Doing so looks challenging to me with the existing admission controllers. Do you have any ideas on addressing this issue?

smarterclayton replied:
Transactional multiobject admission is not possible.

vishh replied:
Would this mean that we have to necessarily have a wrapper (maybe QueueJob) that will allow for admission of a gang of pods in one shot?

liggitt replied:
by the time scheduling happens, the pod objects have already been admitted, which means object counts, resource quotas, and pod spec restrictions have already been enforced... if the gang scheduler is just matching up pods to nodes, it doesn't seem like admission would cause a lot of problems

> Would this mean that we have to necessarily have a wrapper (maybe QueueJob) that will allow for admission of a gang of pods in one shot?

no, you can't end-run around individual pod admission via another object

k82cn replied:
> I think we ideally want to admit either the whole gang or none of the members of a gang. 

It's better to delegate this to controllers or [Applications](https://github.com/kubernetes-sigs/application); it's hard for admit to collect "dynamic info", e.g. for `ResourceQuota`, it's hard for admit to decide which pod should be create to meet 'Gang/PodSchedulingGroup' as much as possible. In google doc, the QueueJob will stop creating new Pod if rejected (there're several corner case to cover)

bsalamat replied:
> In google doc, the QueueJob will stop creating new Pod if rejected (there're several corner case to cover)

The problem is that the user experience would be poor. A few pods of a gang are created, but the rest fail admission and are not created. Those pod created will not be scheduled, but will consume the user's quota. I guess the user has to delete them manually. If we cannot have transactional admission, at lease we should have a controller clean up such pods.

k82cn replied:
> A few pods of a gang are created, but the rest fail admission and are not created. 

For batch or run-to-complete workload, it's better to have some pending jobs there; so k8s can start the pending jobs immediately
when some running jobs finished, instead of waiting for client to submit new jobs; hmm...., a kind of pipeline. The difference is that
the "controller" has to make sure not introduce deadlock (e.g. for every pod group, just start part of them), only the "last" 
pod group should be impacted.

For example, in google doc, the QueueJobController will create pods one by one based on QueueJob's
minAvaialbe; if minAvaialbe pods are not created, it'll NOT create pods for next QueueJob. If all minAvailable are meet,
QueueJobController will create pods up to `.spec.replicas` in round-robin for each QueueJob. 
There're still some cases are not covered, e.g. update minAvailabe for a better SLA, new high priority job,
start other jobs if can not meet minAvailable. That's another reason that prefer to focus on batch/run-to-complete firstly,
and introduce QueueJob; I'm also ok to enhance Job to support those cases in future, and maybe Queue.

Currently, AdmissionController checks pod based on the 'snapshot' of status/config, e.g. ResourceQuota. But for gang scheduling,
it also need to check whether controller **WILL** create more pods later to meet gang requirement. And a wrapper object can not
help as some AdmissionController are still doing pod level check.

Another corner case is that: a hug job (e.g. 1000+ tasks) waits for long running run-to-complete jobs (e.g. hours or days) :(
The candidate solution is to support Queue, and balance the resource sharing between Queue. This's also in backlog.

bsalamat commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+## Function Detail
+
+### CRD Definition
```
I have a slightly different idea about the API. Can we create a **namespaced** object, let's call it a gang, that only specifies properties of a gang. Something like:

```go
type Gang struct {
    metav1.TypeMeta
    metav1.ObjectMeta

    NumMembers int
    Policy GangLifeCyclePolicy
}
```

Then Pod.PodSpec can have a field called `GangName` or `GangId`. When the field is not empty, it means that the pod is gang member. When scheduling pods which are gang members, scheduler makes sure that at least `Gang.NumMember` of such pods can be scheduled simultaneously or none of them are scheduled.

The Gang type can also include gang life cycle policies. For example, the policy can specify that when the number of running pods drops below `NumMember` kill all the other instances, or wait for a certain amount of time for the gang members to reach `NumMember` before killing other instances, ....

Here are some of the benefits of this design:
1. Existing workloads that use one of the K8s collections, e.g. ReplicaSet, Job, etc., can keep their config and only add the `GangName` to their pod template.
1. Flexibility. Pods of a gang can have different PodSpec and belong to different collections.
1. Existing controllers which are responsible for managing life cycle of collections work well with this design.
1. Users who have already learned and worked with K8s collections can keep using the familiar collections.

smarterclayton replied:
This sounds like something that should be a label before we create a field on pod spec (also gang wouldn’t be a great name for an api object).

vishh replied:
We need a way to identify a gang where labels could suffice in theory. We also need to know how many pods comprise a gang (or minimum size of a gang). Encoding that value isn't ideal.
I'd recommend modifying @bsalamat's proposal slightly by including a label selector to identify pods belonging to a gang.

+1 for aiming to have compatibility with existing controllers, Job API at minimum.

k82cn replied:
Used to propose `PodSchedulingGroup` in the [google doc](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit?disco=AAAAB4dD780),
so I totally agree with you to provide such kind of API object in core.

Regarding label selector and 'GangID/GroupID', there're some cases we need to handle:
  - race condition: when a scheduler got a Pod w/o 'Gang/PodSchedulingGroup', how to schedule it? 
    * If schedule the pod and no 'Gang/PodSchedulingGroup' coming, that's OK
    * If schedule the pod, 'Gang/PodSchedulingGroup' coming later and meet 'minAvailable', that's OK
    * If schedule the pod, 'Gang/PodSchedulingGroup' coming later but less than 'minAvailable':
      - if it's because of controller delay, schedule should wait
      - if there's no pod anymore, schedule should handle it by LifeCyclePolicy
    * If does not schedule the pod, how long should we wait?
  - performance: for a batch or run-to-complete workload, the pod come and go in short time; it may take time to
    build relationship between 'Gang/PodSchedulingGroup' and Pod
  - summary/status: if 'Gang/PodSchedulingGroup' support cross Kind/Types, it's better to provide a way
    show the summary/status, such succeeded, running, pending

Here's what I did in kube-arbitrator:
  - race condition: as a batch scheduler, kube-arbitrator do scheduling based on 'job'; so it'll not schedule pods
    until get 'Gang/PodSchedulingGroup'
  - performance: using OwnerReference (controller==true); so 'GangID/GroupID' will be better for performance. 'label selector' is
    kind of short term solution to avoid touching more Kind/Type
  - summary/status: it delegate 'Gang/PodSchedulingGroup' owner to handle summary/status

Another minor comment is the name: as you known, Gang-scheduling is one of requirements, we may enhance this object for other requirements later; so a more
general name maybe better, also including GangLifeCyclePolicy, e.g. changing to PodGroupLifeCyclePolicy. And suppose LifeCyclePolicy is going to be handled
by kube-arbitrator.

ConnorDoyle replied:
+1 to general point raised by @bsalamat here!

For naming, how about something simple like SchedulingGroup? Then the gang-ness can be specified in the policy.

vishh commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+## Motivation
+
+After the discussion at [Gang-scheduling](https://docs.google.com/document/d/1AUwcvTtULNvow5M9e428FnlvINO1uQ7ojRoTGuTp4DA/edit#heading=h.ckn8nv2jj0xv) proposal, we decide to make Gang-scheduling API out of core by CRD, and implement it in [kube-arbitrator](https://github.com/kubernetes-incubator/kube-arbitrator). kube-arbitrator focus on "batch" workload in kubernetes, and will share the same [sc ...
```
I'd recommend focusing on one feature at a time. If that's gang scheduling, just talk about how k8s would enable it either via default scheduler or a new scheduler under incubation.

vishh commented on contributors/design-proposals/scheduling/gang-scheduling.md:
````diff
@@ -0,0 +1,174 @@
+
+```go
+// QueueJob is a Kind for batch workload.
+type QueueJob struct {
````
If you were to consider the approach @bsalamat suggested, we would not need a queue job I think.

k82cn replied:
Yes; if so, I'd like to enhance 'Job' instead of introducing a new Kind.

k82cn replied:
For example, multiple pod template (or wrapper of multiple job).

k82cn commented:
@bsalamat  @vishh  , update the doc according to comments, PTAL :)

1. Introduce `PodSchedulingGroup` into core
1. Added feature interactions section, e.g. Admission

vishh commented:
Hey Klaus, I feel we can move this design further if we have a dedicated
design review over VC in the next few weeks. They're are still quite a few
changes I'd like to see in the design.
Its also not clear what you're proposing to handle lack of transactional
admission of pods.

On Wed, Jul 25, 2018, 11:21 PM Klaus Ma <notifications@github.com> wrote:

> @bsalamat <https://github.com/bsalamat> @vishh <https://github.com/vishh>
> , update the doc according to comments, PTAL :)
>
>    1. Introduce PodSchedulingGroup into core
>    2. Added feature interactions section, e.g. Admission
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/2337#issuecomment-407989535>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKOEn6jz4iniOTAml59b3PD1cTua1ks5uKWAGgaJpZM4VALT2>
> .
>

k82cn commented:
Sure, booked a meeting to sync up :)

-- Klaus


On Fri, Jul 27, 2018 at 1:35 AM Vish Kannan <notifications@github.com>
wrote:

> Hey Klaus, I feel we can move this design further if we have a dedicated
> design review over VC in the next few weeks. They're are still quite a few
> changes I'd like to see in the design.
> Its also not clear what you're proposing to handle lack of transactional
> admission of pods.
>
> On Wed, Jul 25, 2018, 11:21 PM Klaus Ma <notifications@github.com> wrote:
>
> > @bsalamat <https://github.com/bsalamat> @vishh <https://github.com/vishh
> >
> > , update the doc according to comments, PTAL :)
> >
> > 1. Introduce PodSchedulingGroup into core
> > 2. Added feature interactions section, e.g. Admission
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/kubernetes/community/pull/2337#issuecomment-407989535>,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AGvIKOEn6jz4iniOTAml59b3PD1cTua1ks5uKWAGgaJpZM4VALT2
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/community/pull/2337#issuecomment-408175283>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ARRLLRZBm5G-0Yx-B28LXoNCS7j8Jc4Bks5uKf3TgaJpZM4VALT2>
> .
>

ConnorDoyle commented:
Is that meeting open to the public? (Asking for an invite)

k82cn commented:
Here's the link of google calendar :)

https://calendar.google.com/event?action=TEMPLATE&tmeid=NWpwNmdra3EzNmtvOGFlZG9tcjUxcWFvNnVfMjAxODA3MjhUMjMwMDAwWiBrbGF1czE5ODIuY25AbQ&tmsrc=klaus1982.cn%40gmail.com&scp=ALL

k82cn commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,188 @@
+    // MinAvailable defines the minimal available tasks to run the Job;
+    // if there's not enough resources to start all tasks, the scheduler
+    // will not start anyone.
+    MinAvailable int
```
s/MinAvailable/NumMember/g

ConnorDoyle replied:
`NumPods`?

beberg replied:
Minimum? It's likely rare to set less than Replicas, so this should be optional for simplicity.

bsalamat replied:
@beberg it may be needed in many cases for a Job or ReplicaSet, etc. to have more replicas than those needed to be run together.

ConnorDoyle commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,188 @@
+}
+
+// PodSchedulingGroupTemplate represents the template of a pod group.
+type PodSchedulingGroupTemplate struct {
```
s/Template/Spec

k82cn replied:
Done :)

beberg commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,188 @@
+// according to Pod's phase.
+type LifeCyclePolicy struct {
+    // The action that will be taken to the PodSchedulingGroup according to
+    // Pod's phase. One of "Restart", "None".
```
Since the context and point of this API is gang scheduling, "Restart" is the more logical default here for both "PodFailed" and "Unschedulable". The none/none can be done with exiting APIs today. Would also like to know the use case in mind for restart/none and none/restart, as they seem to be non-gang scenarios covered by existing APIs.

bsalamat commented on contributors/design-proposals/scheduling/gang-scheduling.md:
````diff
@@ -0,0 +1,188 @@
+
+```go
+// PodSchedulingGroup defines the scheduling requirement of a pod group
+type PodSchedulingGroup struct {
````
What do you think about renaming this to just `PodGroup`?

k82cn replied:
+1

bsalamat commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,188 @@
+
+    // The phase of pod; the controller takes actions according to this
+    // pod's phase. One of "PodFailed", "Unschedulable".
+    Event Event
```
I think we should also add a Timeout field here. When Timeout is specified, the controller waits for the Group members to satisfy the Group requirements. If the timeout expires, it takes the `Action`.

vishh replied:
What does `PodFailed` mean in this context? Should this field instead be a Phase that indicates the state of the gang/group?

k82cn replied:
> we should also add a Timeout field here.

+1, and Done

> What does `PodFailed` mean in this context? Should this field instead be a Phase that indicates the state of the gang/group?

That means at least one Pod of Group are failed; `PodGroupController` need to take action accordingly. Did not introduce `Phase` for `PodGroup` as only the Pod's "controller" knows phase; for example, for a batch job, there maybe `Failed`, `Done`, `Suspend`, `Terminated` and so on `Phase`; `PodGroupController` does not know that :(.

vishh commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,188 @@
+
+// LifecyclePolicy represents the lifecycle policy of PodSchedulingGroup
+// according to Pod's phase.
+type LifeCyclePolicy struct {
```
As discussed offline, this should be handled by a controller (job++ controller for example). I'd recommend punting lifecycle out of the initial milestone.

k82cn replied:
I keep lifecycle to make this proposal consistent; I added a roadmap section to highlight we're going to handle this in coming release.  The API will add/update accordingly.

vishh commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,188 @@
+
+    // The number of pods which reached phase Failed.
+    // +optional
+    Failed int32
```
Why are we tracking individual pod phases here? This object should only hold the scheduling phase of the gang/group.
If we want to provide easy access to individual pods from a gang, that would be a separate feature right?

k82cn replied:
Yes, I think we can handle this in "controller" proposal :)

vishh commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,188 @@
+    // The name of PodSchedulingGroup that the pod belongs to; the pod
+    // must belong to the PodSchedulingGroup in the same namespace.
+    // +optional
+    GroupName string
```
I assume this will be an alpha field in the API?

k82cn replied:
Yes.

vishh commented:
@k82cn can you post a summary of our recent meeting? Solving admission is key for moving forward with this feature.

k82cn commented:
**Time/Location**: July 31, 2018; Hangout
**Attendee**: Vishnu, Bobby, Conner, Klaus
**Context**:

1. Renamed `Gang` to `PodSchedulingGroup`
   - Using `GroupName` to build relationship between `PodSchedulingGroup` and `Pod`

[vishh]: The name PodSchedulingGroup isn't ideal either since to me it reads as though it's some scheduling "policy" group rather than scheduling a group of mutually dependent pods which is the intent of the API.

[k82cn]: How about `PodGroup` ? As it also includes lifecycle policy right now.

   - Renamed `MinAvailable` to `NumMember`
2. More discussion on Lifecycle management & Admission: what should we do if only part of `PodSchedulingGroup` are created?
   1. Options:
      - Option 1: Added ResourceRequest in `PodSchedulingGroup`, so `ResourceQuotaAdmit` can check whether there's enough quota for the pod group. The con is it's inconvenient to calculate the resource request of `PodSchedulingGroup` which is a collection of pod controlled by other controllers.
      - Option 2: Dedicated controller for new type/kind, e.g. QueueJob; so it can create Pod one by one and manage pod's lifecycle, the con is 1.) that may not support other features, e.g. StatefulSet, 2.) for customized controllers, it still can not resolve the issue. 
      - Option 3: An JSON patch to create Pod in batch; but API server did not support that right now.
      - Option 4: TTL on SchedulingGroup: if the pods are not bound after it expires, delete them all. And this can be done by the controllers too and it does not prevent a race between multiple controllers eyeing for quota at the same time.

    Regarding Option 1:

    [vishh]Bobby and I chatted a bit more about this. We honestly don't see this API being used directly expect other than via controllers which would create a "gang" object along with the required amount of resources after analyzing a "job".

    [k82cn]: hm... , so `PodGroup/Gang` will not cross Kinds, e.g. Job and RC? If so, that is simpler and more reasonable. BTW, does 'cross Kind' is a reasonable requirements?

 
   2. And job dependency makes it worse
 
Some Viewpoints:
 
1. Currently, Admitted Pod may also not be scheduled
2. To avoid such kind of issue, different components, e.g. admission, controller and scheduler, need to work together; but there's no assumption between components right now
3. In batch system, 'job' is the unit, so it can do admission (e.g. Quota) in job level; but in k8s, we did not know resource info until job controller created pods, and k8s does admission in pod (task) level, lost the job context which makes it's hard to re-admission in job level.
4. Quota + namespace (or priority class) defines kind of Queue; it's better to have pending pod here in the 'Queue'; scheduler will balance the resource between 'Queue' to avoid some user/job use more resource than expected. More explanation on this: Quota defines the maximal objects we can create in k8s (to avoid exhaustion); so sum of all quotas should be bigger than total resources (nodes), and the scheduler should help to balance resource between tenant (e.g. by fair-share). In this context, the cluster will have running `PodGroup`, pending `PodGroup`, and few "broken" `PodGroup` (the last group that are impacted by admission). Similar to a queue.
 
Follow up:
 
1. Klaus will check how Slurm/YARN handle that

k82cn commented:
> 1. Klaus will check how Slurm/YARN handle that

Checked Slurm, it provides **job** level validation for submission and update ([doc](https://slurm.schedmd.com/job_submit_plugins.html)). I'll confirm with Slurm community for detail.

k82cn commented:
> Solving admission is key for moving forward with this feature.

Another option is that: If  `PodGroup` unschedulable, restart it forcely after grace period expired; and add a backoff ratio admission, e.g. reject pod creation after another `Timeout` (or other config).

liggitt commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,213 @@
+    // The name of PodGroup that the pod belongs to; the pod
+    // must belong to the PodGroup in the same namespace.
+    // +optional
+    GroupName string
```
[feedback from api reviewers](https://github.com/kubernetes/community/pull/2337#discussion_r203909088) was that this could/should be modeled as a label rather than a spec field. I don't understand the point about a race condition raised in https://github.com/kubernetes/community/pull/2337#discussion_r203954962... a `metadata.labels['scheduling.k8s.io/group-name']` label on a pod carries just as much information as a `spec.groupName` field, and has the additional benefit of allowing label-selection queries to filter pods to a specific group name

vishh replied:
@liggitt 
I like the flexibility of this new pattern, but aren't system components supposed to not look for specific label keys for their core functionalities?

liggitt replied:
I thought the scheduler acting on this was being built out of tree

k82cn replied:
Done :)

liggitt commented on contributors/design-proposals/scheduling/gang-scheduling.md:
````diff
@@ -0,0 +1,213 @@
+} 
+```
+
+The `.spec.GroupName` specifies the `PodGroup` that it belongs to; and the pod can only belong to the `PodGroup` in the same namespace. The pod, controlled by different collections, can also belong to the same `PodGroup`.  Because of performance concern,  it does not use `LabelSelector` to build the relationship between `PodGroup` and `Pod`.
````
> Because of performance concern,  it does not use `LabelSelector` to build the relationship between `PodGroup` and `Pod`.

What is the difference, performance-wise, between pulling the group name out of a label, and pulling it out of a spec field?

vishh replied:
+1 Performance is not the main issue. Are we relaxing [rules on labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) - specifically,
"Labels are intended to be used to specify identifying attributes of objects that are meaningful and relevant to users, but do not directly imply semantics to the core system."

k82cn replied:
The performance is about a selector in PodGroup; for run-to-complete workload, the status will be changed frequently; and there'll be more pending job here.

For the label, maybe annotation for alpha features as before. We do not need selector for apiserver, as scheduler will watch all pods; it need a way to build the relationship between PodGroup and Pod quickly.

BTW, the validation of `metadata.name` and `label` seems different in my mind.

bsalamat replied:
There is a big performance difference between label selectors in a PodGroup and a field in PodSpec. If we allow label selectors in PodGroups to choose pods, we would need to match the label selector of all PodGroups against any pod being scheduled to find whether the Pod is a member of the group or not. It also has the problem that a Pod may end up in multiple PodGroups. Then we would need to have some sort of priority in PodGroups to pick one.
Moreover, when any pod terminates we would again need to match the label selectors of all PodGroups against the pod to find if it was a member of a group. If so, we would need to match the selector against all running pods to determine whether the number of member has dropped below the allowed value or not.

tallclair commented on contributors/design-proposals/scheduling/gang-scheduling.md:
````diff
@@ -0,0 +1,203 @@
+scheduling.k8s.io/group-name
+```
+
+The `scheduling.k8s.io/group-name` annotation specifies the `PodGroup` that it belongs to; and the pod can only belong to the `PodGroup` in the same namespace. The pod, controlled by different collections, can also belong to the same `PodGroup`.  Because of performance concern,  it does not use `LabelSelector` to build the relationship between `PodGroup` and `Pod`.
````
Perhaps I'm missing something, but how does this support pods with different specs in the same group? Specifically, suppose I have 2 ReplicaSets, "foo" and "bar". I want foo pods & bar pods to be scheduled together, in a PodGroup. Is there any way to guarantee that the group will consist of a foo & bar pod, and not 2 foos, or 2 bars?

k82cn replied:
> but how does this support pods with different specs in the same group? 

By other algorithm plugins to work together with gang-scheduling.

> Is there any way to guarantee that the group will consist of a foo & bar pod, and not 2 foos, or 2 bars?

What's case of that? Currently, a job may include master and worker tasks; so there's TaskPriority plugin in kube-arbitrator to make sure master tasks started firstly in PodGroup.

bsalamat replied:
@tallclair Gang scheduling addresses the problem of scheduling a number of pods together. For example, you may want to schedule all members of two ReplicaSets together. Gang scheduling is not addressing pod affinity or pod dependency issues.
In the case of running all pods in multiple collections together, one can provide the name of one PodGroup in their spec/annotation. The PodGroup.Spec.NumMembers is then set to the sum of number of pods in the two collections. This configuration guarantees that they will be scheduled together.

k82cn commented:
> I think discovery is a critical aspect of most pod group use cases, but it's not mentioned in this proposal.
> Specifically if 2 pods are grouped together, they likely want to communicate with each other, and should be able to pick the correct instance.

I'd like to handle this by an algorithm plugin of scheduler to do that :)

> This looks a bit like "MetaPods"

What's "MetaPods"? Can you share the link?

tallclair commented:
> What's "MetaPods"? Can you share the link?

It's an idea we tossed around in regards to the [sandbox isolation decision](https://docs.google.com/document/d/1fe7lQUjYKR0cijRmSbH_y0_l3CYPkwtQa5ViywuNo8Q/edit#heading=h.y2rkt6glu7p6). Basically, if you want 2 pods to behave like 1 pod, but have stronger isolation between them. Gang scheduling + discovery + lifecycle management.

bsalamat commented:
On a slightly different note, I think we should be able to provide a member ID to each member of a PodGroup, for example, via an environment variable. Such member ID will be useful in many data processing pipelines. Each member uses this ID to access a chunk of data that is assigned to it. This could have many more applications, in pod groups where certain members communicate with an assigned pair, or parent (in tree-like hierarchies).

k82cn commented:
> On a slightly different note, I think we should be able to provide a member ID to each member of a PodGroup, for example, via an environment variable. 

Similar to Indexed Job (https://github.com/kubernetes/kubernetes/issues/14188) ? That's required by some workload, e.g. MPI; personally, I'd like to handle it in a seperate feature :)

bsalamat commented:
> Similar to Indexed Job (kubernetes/kubernetes#14188) ?

Yes

> That's required by some workloads, e.g. MPI; personally, I'd like to handle it in a seperate feature :)

I agree. It should be feature for one of our collections (Job, etc).

k82cn commented:
@vishh , @bsalamat , @liggitt , do you have more comments ?

Regarding the "transaction" of pod group, I'm working on the prototype of Queue for more pending pods to give a toleration of transaction. And I think it can be a seperate discussion to follow up :)

liyinan926 commented on contributors/design-proposals/scheduling/gang-scheduling.md:
````diff
@@ -0,0 +1,203 @@
+
+```go
+// PodGroup defines the scheduling requirement of a pod group
+type PodGroup struct {
````
How will this definition meet the requirement `2. Pods of a group/gang may have different `PodSpec` (and/or belong to different collections)`? Will this supports scheduler more than one groups together?

k82cn replied:
Regarding multiple/different PodSpec, user can use multiple job or customized job with multiple podSpec; the requirement is to have same GroupName. here's an example ([QueueJob](https://github.com/kubernetes-incubator/kube-arbitrator/blob/master/pkg/apis/extensions/v1alpha1/job.go#L54)) in kube-arbitrator.

And we also support `TaskPriority` within Job to make sure driver will be started before executor, also preemption within Job. There's another topic about how driver/executor find each other if using gang-scheduling + job; will propose something later, or if you have some suggestion.

vishh commented:
@k82cn I'd like to see a practical design for handling transactional admission. I feel introducing a queue adds more complexity. There will be queues for other purposes like budgeting, prioritization, etc., and how will such a queue interact with a gang queue?
Have you explored the other idea I had proposed - secure resources for the gang first and then admit pods belonging to that gang?

tanshanshan commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,203 @@
+
+Usually, CRD ([CustomResourceDefinitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)) feature is used to introduce a customized **Kind**, named `CRDJob` as example. The customized controller, named `CRDJobController`, watches it and manages lifecycle of it:
+
+1. For each `CRDJob`, `CRDJobController` creates a `CRDJob` and `PodGroup` (one `CRDJob` with one `PodGroup` as example). The attributes of `PodGroup`should be set accordingly, e.g `numMember`; it's up to customized controller on how to manage relationship between `PodGroup` and `CRDJobob`, e.g. `metadata.name`.
```
s/CRDJobob/CRDJob

k82cn replied:
done, thanks :)

tanshanshan commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,203 @@
+Usually, CRD ([CustomResourceDefinitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/)) feature is used to introduce a customized **Kind**, named `CRDJob` as example. The customized controller, named `CRDJobController`, watches it and manages lifecycle of it:
+
+1. For each `CRDJob`, `CRDJobController` creates a `CRDJob` and `PodGroup` (one `CRDJob` with one `PodGroup` as example). The attributes of `PodGroup`should be set accordingly, e.g `numMember`; it's up to customized controller on how to manage relationship between `PodGroup` and `CRDJobob`, e.g. `metadata.name`.
+2. When `CRDJobobController` create Pods, its annotation should be set accordingly. `kube-arbitrator` follows gang-scheduling logic to schedule those pods in batch.
```
s/CRDJobobController/CRDJobController

k82cn replied:
done, thanks :)

tanshanshan commented on contributors/design-proposals/scheduling/gang-scheduling.md:
```diff
@@ -0,0 +1,203 @@
+
+### Admission Controller
+
+Because of AdmissionController, e.g. `ResourceQuota`, a few pods of a gang maybe created, but the rest fail admission pods are not created. For batch or run-to-complete workload, it's better to have some pending jobs there; so it can start the pending jobs immediately when some running jobs finished, instead of waiting for client to submit new jobs; a kind of pipeline. The difference is that the  ...
```
s/ batch or run-to-complete/  batch/run-to-complete

k82cn commented:
> @k82cn I'd like to see a practical design for handling transactional admission. I feel introducing a queue adds more complexity. 
> There will be queues for other purposes like budgeting, prioritization, etc.,

Yes, there're several feature for queue; but in this context, I'd like to leverage pending and sharing capability of queue:
  - pending task/'job': because of queue, controllers can create more pods than cluster resources; and controllers should try to create
    pods of jobs in serial. That reduces restrictions on transaction. And because of lifecycle of PodGroup, the PodGroup will be restart
    if not enough pod for a while
  - sharing: because of pending task/'job', we need sharing capability to avoid part of users exhaust the whole cluster

> and how will such a queue interact with a gang queue?

We support gang-scheduling within a Queue; currently, I map Queue to namespace with some additional attributes, e.g. weight for sharing.

> Have you explored the other idea I had proposed - secure resources for the gang first and then admit pods belonging to that gang?

Oh, there's some case in my mind we need to handle for this proposal:

  - there maybe hundreds of tasks (pod) for a job; apiserver will hold hundreds of connection for one job until we reject or accept all creations
  - if there're some PodGroups created in parallel, we need an algorithm in admission controller to decide which PodGroup should be admit firstly, e.g. by priority
  - as we discussed before, admitted pods may also not be scheduled because of predicates; we can not leverage other advanced feature to enhance it, e.g. backfill

So I think queue is better than handling transaction in admission

k82cn commented:
@vishh , any more comments ?

k82cn commented:
@vishh , @bsalamat , did the following update to this doc according to the discussion:

 1. Removed the lifecycle management section from this proposal, but kube-arbitrator will record event for controller, e.g. unschedulable.
 1. Removed the discussion about Queue
 1. Added `Resources` field into `PodGroupSpec` for quota

If anymore comments for this doc, please let me know :)

I'll bring other doc/brainstorming thread for the other topics, e.g. controller of PodGroup, Queue, and batch scheduler + default scheduler.

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+`kube-arbitrator` only watches `PodGroup` and `Pod`. It'll reconstruct 'Job' by annotation of Pod and `PodGroup`, the `Pod`s are considered as 'Task' of 'Job'; if annotation is empty, `kube-arbitrator` records an unschedulable event of pod to ask user/controller to resubmit it to `kube-scheduler`. `kube-arbitrator` does not schedule pods until its `PodGroup` is ready.
+
+As `kube-arbitrator` and `kube-scheduler` maybe running in parallel; `kube-arbitrator` follows multi-scheduler feature to only handle the `Pod` that submitted to `kube-arbitrator`. The `kube-arbitrator` does scheduling as follow:
```
s/maybe/may be/

and a few other cases below.

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+As `kube-arbitrator` and `kube-scheduler` maybe running in parallel; `kube-arbitrator` follows multi-scheduler feature to only handle the `Pod` that submitted to `kube-arbitrator`. The `kube-arbitrator` does scheduling as follow:
+
+1. Reconstruct 'Job' by the annotation of `Pod` and `PodGroup`
+2. If `PodGroup` is not ready, the 'job' will not be scheduled; and an unschedulable event of `Pod` will be recorded
```
You should explain what a PodGroup not being ready means.

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+1. Reconstruct 'Job' by the annotation of `Pod` and `PodGroup`
+2. If `PodGroup` is not ready, the 'job' will not be scheduled; and an unschedulable event of `Pod` will be recorded
+3. In `allocate` phase, `kube-arbitrator` will
+   * record an `Unschedulable` event of `PodGroup` if some pods are running but `succeeded + pending + running < minAvailable`, the `PodGroupController` takes action according to `LifecyclePolicy`
```
You have removed `LifecyclePolicy` from the API.

bsalamat replied:
s/`minAvailable`/`numMembers`/

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+    // NumMember defines the number of members/tasks to run the pod group;
+    // if there's not enough resources to start all tasks, the scheduler
+    // will not start anyone.
+    NumMember int
```
s/NumMember/NumMembers/

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+3. In `allocate` phase, `kube-arbitrator` will
+   * record an `Unschedulable` event of `PodGroup` if some pods are running but `succeeded + pending + running < minAvailable`, the `PodGroupController` takes action according to `LifecyclePolicy`
+   * allocate (but not bind) resource to Pods according to Pod's spec, e.g. `NodeAffinity`
+   * bind all Pods to hosts until job ready (`numMember` <= `allocated Pods` + `succeeded Pods`)
```
There may be use-cases for keeping a minimum number of pods running, no matter how many of them have succeeded. For example, I may have 1000 pods. I want to run at least 10 of them together regardless of how many have already run and finished. Of course, the last batch of them is an exception and it is ok if the number of the last batch is less than 10. For this reason, as long as `numMembers` <= `allocated Pods` + `pending Pods`, we should bind only when `numMembers` <= `allocated Pods`. When there are not enough pending pods to satisfy `numMembers` <= `allocated Pods` + `pending Pods`, then we can bind when `numMember` <= `allocated Pods` + `succeeded Pods`.

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+   * record an `Unschedulable` event of `PodGroup` if some pods are running but `succeeded + pending + running < minAvailable`, the `PodGroupController` takes action according to `LifecyclePolicy`
+   * allocate (but not bind) resource to Pods according to Pod's spec, e.g. `NodeAffinity`
+   * bind all Pods to hosts until job ready (`numMember` <= `allocated Pods` + `succeeded Pods`)
+4. If can not allocate enough resources to the job, the pods are keeping pending; and the resource can not allocate to other job
```
s/pods are keeping pending/pods stay pending/
s/resource can not allocate/resources cannot be allocated/

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+   * record an `Unschedulable` event of `PodGroup` if some pods are running but `succeeded + pending + running < minAvailable`, the `PodGroupController` takes action according to `LifecyclePolicy`
+   * allocate (but not bind) resource to Pods according to Pod's spec, e.g. `NodeAffinity`
+   * bind all Pods to hosts until job ready (`numMember` <= `allocated Pods` + `succeeded Pods`)
+4. If can not allocate enough resources to the job, the pods are keeping pending; and the resource can not allocate to other job
```
Should we have a timeout for keeping member of a PodGroup in "allocated" phase? We shouldn't block some of the cluster resources indefinitely.

k82cn replied:
hm... we do not have timeout for now; I'd like to use 'backfill' to avoid idle resources.

vishh replied:
If this is the logic within the scheduler I agree with @bsalamat that resources should be held for too long if a pod group cannot be scheduled.
From a user POV, they should be able to set `activeDeadlineSeconds` to their application specification to ensure that podgroups that are unschedulable do not end up hogging system resources for too long.

k82cn replied:
Currently, kube-arbitrator does scheduling every 1 second; in each scheduling cycle, it'll try to allocate resources to the job. If there's not enough resource to start a job for Coscheduling, the resource is not allocated to others right now; so I'd like to add a phase to backfill smaller job to improve the utilization later. Will not consider expected execution time at the beginning, preempt the backfill-ed job if "high" priority job can run it.

k82cn replied:
btw, for this kind of backfill, we only handle short running job firstly, e.g. streaming task. for expected execution time, may consider it later.

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+1. For each `CRDJob`, `CRDJobController` creates a `CRDJob` and `PodGroup` (one `CRDJob` with one `PodGroup` as example). The attributes of `PodGroup`should be set accordingly, e.g `numMember`; it's up to customized controller on how to manage relationship between `PodGroup` and `CRDJob`, e.g. `metadata.name`.
+2. When `CRDJobController` create Pods, its annotation should be set accordingly. `kube-arbitrator` follows gang-scheduling logic to schedule those pods in batch.
+3. When pods failed/deleted/unschedulable, it up to `CRDJobController` on how to manage `CRDJob`'s lifecycle. For example, if `CRDJobController` manages lifecycle itself, set `.spec.Policy` of `PodGroup` to nil; otherwise, `PodGroupController` will manage the lifecycle as described above.
```
s/it up/it is up/

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+### Multi-scheduler
+
+When multi-scheduler was enabled, there maybe decision conflict between different schedulers; and the kubelet will reject one pod (failed) if conflict. The controller will handle rejected pods based on its lifecycle policy for failed pods. It up to cluster admin to avoid such kind of conflict, e.g. node affinity which is supported by kube-arbitrator in 1.13.
```
s/When multi-scheduler was enabled/Since multiple schedulers work in parallel/

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+### Priority/Preemption
+
+As described in `multi-scheduler` section, it's better to run batch and service workload in different zone to avoid confliction.
```
s/confliction/conflicts/

Here and below

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+### Priority/Preemption
+
+As described in `multi-scheduler` section, it's better to run batch and service workload in different zone to avoid confliction.
+Rejected a batch/run-to-complete pod may trigger the restarting of the whole `PodGroup` which impact the performance. The solution on how to handle confliction better will be proposed in coming release.
```
s/Rejected a batch/run-to-complete pod/A preempted batch/run-to-completion Pod/

k82cn replied:
hm.. I think it should be "rejected" which means rejected by kubelet.

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+### Pod RestartPolicy
+
+Pod's `RestartPolicy` still works as before. But for batch/run-to-compelete workload, it's better to set `RestartPolicy` to `Never` to avoid endless restart loop.
```
Why is batch distinguished here? Restart policy could similarly cause endless restart loop in other types of tasks. Kubelet has backoff mechanism to avoid a tight restart loop.

vishh replied:
I feel this tip is not relevant to this proposal. It is better suited for a controller design proposal/user-guide. And the job controller already has some [documentation on this](https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/#handling-pod-and-container-failures)

k82cn replied:
several people asked this when using kube-arbitrator, so I highlight here. I'm ok to highlight in readme or other place.

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+### Admission Controller
+
+Because of AdmissionController, e.g. `ResourceQuota`, a few pods of a gang maybe created, but the rest fail admission pods are not created. `.spec.Resource` is introduced in `PodGroup` to reserve resources for PodGroup by `ResourceQuota`, so it makes sure the Pods of PodGroup has enough quota to create it.
```
`spec.Resource` is specifically added to address the `ResourceQuota` admission check. `spec.Resource` is not enough to address all challenges of admitting batches of pods. So, I would rephrase this as:

If quota runs out in the middle of creating a group of pods, a few members of a PodGroup may be created, while the rest will be denied by the `ResourceQuota` admission controller. `.spec.Resource` is added in `PodGroup` to address this problem. When a `PodGroup` is created with `.spec.Resource`, so much quota is reserved for the group if there is available quota. Pods of group use the already reserved quota. By setting `.spec.Resource` properly, one can ensure that Pods of a PodGroup have enough quota at creation time.

vishh replied:
What happens if a pod has to be vertically scaled up (this is a WIP AFAIK)? I'd assume the podGroup has to be scaled up first successfully, before attempting to scale up pods in it. Will that be supported?

vishh replied:
And will scale down's be supported at all?

k82cn commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+### Lifecycle Management
+
+As the lifecycle of Pods in PodGroup maybe different from controller to controller, the lifecycle management is delegated to the owner of Pods. In kube-arbitrator, it'll record related events for controller to manage pods, e.g. `Unschedulable`.
```
> We probably don't need to address this in the first version, but we might want to revisit this decision later.

Yes. I'd like to have a controller as default implementation to manage `PodGroup` as it may cover most of case, e.g. restart the whole PodGroup. I'll open the other PR for it :)

vishh replied:
It is not clear what kube-arbitrator is in this context. Can I suggest dropping that from this proposal since this is primarily about the scheduler?
+1 on having a separate conversation about a controller with built in podgroup support and lifecycle management primitives. I feel that can be done without having to go through the heavy weight process of a KEP.

k82cn replied:
> It is not clear what kube-arbitrator is in this context. Can I suggest dropping that from this proposal since this is primarily about the scheduler?

OK, I'll clarify it's batch scheduler to do that. And add kube-arbitrator into implementing plan.

> I feel that can be done without having to go through the heavy weight process of a KEP.

That's great. As PodGroup is CRD, I'd like to build this controller into kube-arbitrator which's related.

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+    // Resources defines the total resource the PodGroup requests to run
+    // Pods.
+    Resources v1.ResourceList
```
+1

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+type PodGroupStatus struct {
+    // The number of actively running pods.
+    // +optional
+    Running int32
```
Can we add a list of `Admitted` pods as well? Rationale is that once a pod group resource quota request is granted, we want to know how many pods have already been admitted as part of the group. 

Also, would it make sense to have a list of local object references to identify pods rather than just an integer count. That way it will be easier for higher level controllers to track the phases of pods belonging to a pod group easily.

k82cn replied:
> Also, would it make sense to have a list of local object references to identify pods rather than just an integer count.

Wanted to add such detail for user; but I think the customized controller will provide such info. Prefer to keep it simple right now, if it's necessary, let's update it in other PRs. That's the benefits we got from CRD :)

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
````diff
@@ -0,0 +1,174 @@
+}
+```
+
+The `PodGroup`, which is a namespaced object, specifies the behavior and status of a pod group, e.g. number of pods in a group. To define which pods are member of `PodGroup`, the following annotation key is introduced for `Pod`.
````
nit: `number of pods` is not a `behavior`.

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
````diff
@@ -0,0 +1,174 @@
+scheduling.k8s.io/group-name
+```
+
+The `scheduling.k8s.io/group-name` annotation specifies the `PodGroup` that it belongs to; and the pod can only belong to the `PodGroup` in the same namespace. The pod, controlled by different collections, can also belong to the same `PodGroup`.  Because of performance concern,  it does not use `LabelSelector` to build the relationship between `PodGroup` and `Pod`.
````
In that case, why not consider adding an alpha field?

bsalamat replied:
An alpha field should be added to core, but PodGroup will be a CRD. So, I think we should postpone adding the field to sometime in the future when we move PodGroup to the core.

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+As the lifecycle of Pods in PodGroup maybe different from controller to controller, the lifecycle management is delegated to the owner of Pods. In kube-arbitrator, it'll record related events for controller to manage pods, e.g. `Unschedulable`.
+
+The update to `PodGroup` is not support for now; and deleting `PodGroup` does not impact Pod's status.
```
What is the lifecycle of a `PodGroup`? Does it have a `Phase`? When is a PodGroup ready to admit pods? What happens when a pod is re-created after it's podGroup is deleted?

k82cn replied:
> What is the lifecycle of a PodGroup? Does it have a Phase?

Did not introduce phase for `PodGroup` right now; prefer to let `Kind` (e.g. Job) to manage that; `PodGroup` is more about scheduling part. 

> What happens when a pod is re-created after it's podGroup is deleted?

If `podGroup` is deleted, only Quota will take effect and batch scheduler will not schedule new pending pod
belong to that `PodGroup`. When we want to support some sharing policy, we may give a 'flag' to pod that
it can be scheduled by batch scheduler with `PodGroup`.

vishh replied:
> Did not introduce phase for PodGroup right now

How will we differentiate between a PodGroup that's pending and one that's been admitted? When is it ok to GC pod groups?

vishh replied:
> If podGroup is deleted, only Quota will take effect

If podgroup get's deleted, will the quota controller automatically take into account the requests & limits of pods that had previously belonged to the podgroup? I assume the answer is yes, but state it explicitly which would make implementation easier.

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+The update to `PodGroup` is not support for now; and deleting `PodGroup` does not impact Pod's status.
+
+### kube-arbitrator
```
What is kube-arbitrator?

vishh replied:
From reading below I see `kube-arbitrator` means an alternate scheduler in this construct. Given that there is general agreement to promote this feature to default scheduler in the future, should it be added to the default scheduler as an experimental (disabled by default) feature now?

k82cn replied:
> should it be added to the default scheduler as an experimental (disabled by default) feature now?

We'd like to use "scheduler framework" to handle that; and Coscheduling/Gang-scheduling just one requirements for batch workload. I'd like to have a separated project to introduce "job" concept in scheduler (to work with scheduler framework) and collect enough requirements.

BTW, we incubate `kube-arbitrator` because of complexity of "batch" :)

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,174 @@
+
+### Kubectl
+
+kubectl is enhanced to support `PodGroup`, including its status.
```
If PodGroup is a CRD, perhaps [kubectl plugins](https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/) can be used to begin with?

k82cn replied:
Yes.

timothysc commented on contributors/design-proposals/scheduling/coscheduling.md:
````diff
@@ -0,0 +1,174 @@
+
+```go
+// PodGroup defines the scheduling requirement of a pod group
+type PodGroup struct {
````
Just being honest, but I think this breaks the fundamental axiom of a pod itself.  The granularity of co-scheduling is done w/multiple containers via the pod.  Co-scheduling an entire pod-group is almost antithetical to the pod abstraction itself.

bsalamat replied:
I objectively disagree. Pod ensures that all of the containers within a pod are scheduled together on a "single node". Coscheduling (or a PodGroup) does not change that concept at all. Pods in a PodGroup do not necessarily land on a single node and even their lifecycle is not necessarily managed together.

timothysc replied:
Then I'd reconsider the order of the content to flow from concept A->B making the case clear.  The init read flow forces the reader to read the entire doc on multiple passes in order to make sense out of the exact details.

vishh replied:
This proposal lacks a proper description of the feature which should be fixed @k82cn.
+1 to @bsalamat's point that this feature is not related to the concept of pods at all. It is a means to represent "mutually dependent pods" that have both scheduling and lifecycle constraints.

k82cn replied:
done, added some background in motivation :)

k82cn commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+    // NumMembers defines the number of members/tasks to run the pod group;
+    // if there's not enough resources to start all tasks, the scheduler
+    // will not start anyone.
+    NumMembers int
```
I'm thinking to rename this field to `minMember`, as there're maybe more members (Pod) than this number; several user confuse about this number with total number of pods :)

The Pod more than this number are scheduled FCFS by default, although there's an experimental [DRF](https://cs.stanford.edu/~matei/papers/2011/nsdi_drf.pdf) policy which is simple enough.

@bsalamat , @vishh , @ConnorDoyle :)

tallclair commented:
/uncc

liggitt commented:
/uncc

k82cn commented:
@bsalamat , @vishh do you have more comments ? :)

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+
+As the lifecycle of Pods in PodGroup may be different from controller to another, the lifecycle of the members is not managed by the coscheduling feature. Each collection controller may implement or already have the mean to manage lifecycle of its members. The scheduler'll record related events for controller to manage pods, e.g. `Unschedulable`.
+
+The update to `PodGroup` is not support for now; and deleting `PodGroup` does not impact Pod's status.
```
nit: s/support/supported/

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+As batch scheduler and default scheduler may be running in parallel; the batch scheduler follows multi-scheduler feature to only handle the `Pod` that submitted to it. The batch scheduler does scheduling as follow:
+
+1. Reconstruct 'Job' by the annotation of `Pod` and `PodGroup`
+2. If there are less Pods than `numMembers` of `PodGroup`, the 'job' will not be scheduled; and an unschedulable event of `Pod` will be recorded
```
please replace numMembers with minMembers here and below.

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+
+As batch scheduler and default scheduler may be running in parallel; the batch scheduler follows multi-scheduler feature to only handle the `Pod` that submitted to it. The batch scheduler does scheduling as follow:
+
+1. Reconstruct 'Job' by the annotation of `Pod` and `PodGroup`
```
It is not clear what this item means. What do you mean by "reconstructing job"?

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
````diff
@@ -0,0 +1,180 @@
+}
+```
+
+The `PodGroup`, which is a namespaced object, specifies the attributes and status of a pod group, e.g. number of pods in a group. To define which pods are member of `PodGroup`, the following annotation key is introduced for `Pod`; the annotation key is used for this alpha feature, and it'll be changed to a more permanent form, such a field, when moing `PodGroup` to core.
````
s/moing/moving/

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+1. Reconstruct 'Job' by the annotation of `Pod` and `PodGroup`
+2. If there are less Pods than `numMembers` of `PodGroup`, the 'job' will not be scheduled; and an unschedulable event of `Pod` will be recorded
+3. In `allocate` phase, scheduler will
+   * record an `Unschedulable` event of `PodGroup` if some pods are running but `succeeded + pending + running < numMembers`, the controller takes action according to its configuration
```
Today, we do not delete completed pods as long as their parent object (such as Job, ReplicaSet, etc) is not deleted. This behavior is essential for the proposed algorithm to work correctly.

(This is a note to ourselves. Nothing needs to be changed in the design.)

vishh replied:
Is the `minMembers` field in PodGroup expected to track the number of non-terminated pods or all pods that have been successfully admitted?

bsalamat replied:
Vish, I think "succeeded" means those pods that have run and finished successfully. This makes sense for "run-to-completion" pods.

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+
+### Multi-scheduler
+
+Since multiple schedulers work in parallel, there may be decision conflict between different schedulers; and the kubelet will reject one pod (failed) if conflict. The controller will handle rejected pods based on its lifecycle policy for failed pods. It up to cluster admin to avoid such kind of conflict, e.g. node affinity which is supported by kube-arbitrator in 1.13.
```
I would change the last sentence to
"Users and cluster admins may reduce the probability of such conflicts by partitioning the clusters logically, for example, by placing node-affinity to distinct set of nodes on various groups of pods."

bsalamat commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+
+The `scheduling.k8s.io/group-name` annotation specifies the `PodGroup` that it belongs to; and the pod can only belong to the `PodGroup` in the same namespace. The pod, controlled by different collections, can also belong to the same `PodGroup`.  Because of performance concern,  it does not use `LabelSelector` to build the relationship between `PodGroup` and `Pod`.
+
+### Lifecycle Management
```
We need to explain whether we take any action when the number of running pods drop below `PodGroup.MinMembers` at run-time.

k82cn replied:
Yes, includes this as part of PodGroup controller.

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+2. Pods of a group/gang may have different `PodSpec` (and/or belong to different collections)
+3. Existing controllers which are responsible for managing life cycle of collections work well with this feature
+
+To meet the requirements above, the following **Kind** is introduced by CRD under `scheduling.incubator.k8s.io/v1alpha1` **Group**/**Version**.
```
nit: Should it be `incubator.scheduling.k8s.io`? That way once it's out of incubation we only need to drop a sub-domain?

k82cn replied:
Yours seems better :)

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+
+    // The number of pods which reached phase Failed.
+    // +optional
+    Failed int32
```
Why is tracking `Succeeded` and `Failed` useful at the pod Group level? This seems to make an assumption that pod groups are typically used only for run to completion jobs.

k82cn replied:
Just give a summary of pod status here; not only for completion jobs.

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
````diff
@@ -0,0 +1,180 @@
+The `PodGroup`, which is a namespaced object, specifies the attributes and status of a pod group, e.g. number of pods in a group. To define which pods are member of `PodGroup`, the following annotation key is introduced for `Pod`; the annotation key is used for this alpha feature, and it'll be changed to a more permanent form, such a field, when moing `PodGroup` to core.
+
+```go
+scheduling.k8s.io/group-name
````
nit: Mark it alpha explicitly & be consistent with the object kind - `alpha.incubator.scheduling.k8s.io`

k82cn replied:
hm .... , mentioned this alpha feature in doc and release notes (in future); so when we migrate this feature to upstream, the user did not need update annotation in pod.

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+
+### Scheduling
+
+The scheduler only watches `PodGroup` and `Pod`. It'll reconstruct 'Job' by annotation of Pod and `PodGroup`, the `Pod`s are considered as 'Task' of 'Job'; if annotation is empty, the scheduler records an unschedulable event of pod to ask user/controller to resubmit it. The schduler does not schedule pods until its `PodGroup` is created.
```
> if annotation is empty, the scheduler records an unschedulable

So the batch scheduler that you intend to develop will not schedule pods that do not need gang scheduling?

k82cn replied:
Yes, for the first version. As we do not when/whether scheduler will get `PodGroup` in future, the `PodGroup` is required currently, but user can set `minMember` to 1.

I'm also thinking to add another "flag" to identify whether `PodGroup` is required to schedule the pod. Supporting jobs without gang-scheduling in kube-batch should be another proposal.

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+As batch scheduler and default scheduler may be running in parallel; the batch scheduler follows multi-scheduler feature to only handle the `Pod` that submitted to it. The batch scheduler does scheduling as follow:
+
+1. Reconstruct 'Job' by the annotation of `Pod` and `PodGroup`
+2. If there are less Pods than `numMembers` of `PodGroup`, the 'job' will not be scheduled; and an unschedulable event of `Pod` will be recorded
```
implementation note: Do not overwhelm the system with repeated unschedulable events.

k82cn replied:
sure :)

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+   * record an `Unschedulable` event of `PodGroup` if some pods are running but `succeeded + pending + running < numMembers`, the controller takes action according to its configuration
+   * allocate (but not bind) resource to Pods according to Pod's spec, e.g. `NodeAffinity`
+   * bind all Pods to hosts until job is ready: if `numMembers` <= `allocated Pods` + `pending Pods`, it's ready when `numMembers` <= `allocated Pods`; otherwise, `numMember` <= `allocated Pods` + `succeeded Pods`
+4. If can not allocate enough resources to the job, the pods stay pending; and the resource cannot be allocated to other job
```
Two issues to consider:
1. There is the risk of split brain where batch scheduler and the default scheduler might schedule onto the same resources which could in theory lead to one of the co-scheduled pods being rejected by the kubelet. What would you do if the batch scheduler cannot re-schedule that pod? Do we fall back on the controller to kill & resubmit the entire job? That feels like a sub-optimal solution from a user POV. Would it be possible to assign phantom pods to nodes to hold on to resources?
2. Deadlock in the batch scheduler - If there are two jobs and both of them have half of their jobs allocated (not bound), then how will you make progress? Have you considered doing a transactional allocation - allocate all or free up allocated resources?

k82cn replied:
> Do we fall back on the controller to kill & resubmit the entire job? That feels like a sub-optimal solution from a user POV. 

Yes, controller will re-submit the entire job. phantom pods maybe also preempted by high priority pods from default scheduler. For such a mix environment, prefer to only share resources between elastic 'jobs', e.g. Spark's executor and nginx/tomcat.

> Deadlock in the batch scheduler ... allocate all or free up allocated resources?

In kube-batch 0.1, we use this option; two major concerns: 1. there're several computing here,  pre-allocated and free up, 2. after free up, the resource may not be used by others, e.g. predicates.
So in kube-batch 0.2/0.3, I'm thinking to use preemption (or backfill) to handle this case: kube-batch will try to allocate resource by jobs's `minMember` as much as possible; if not enough resource, preempt `allocated` resource firstly which is occupied by pod-group but can not use it.

bsalamat replied:
> So in kube-batch 0.2/0.3, I'm thinking to use preemption (or backfill) to handle this case: kube-batch will try to allocate resource by jobs's minMember as much as possible; if not enough resource, preempt allocated resource firstly which is occupied by pod-group but can not use it.

What triggers this "preemption"? When there are two pod groups with the same priority and each one is partially allocated and there is no more resources in the cluster, which one preempts the other?

k82cn replied:
not sure which term is better: preempt or backfill; anyway, the assumed/allocated resource will be in "Allocated" state if the job did not get enough resource; and after allocation phase, those allocated but not bound resource will be re-calculated for other jobs. For your case, the two jobs will try to get those allocated resources in order (e.g. FCFS).

bsalamat replied:
I am not sure if FCFS works in this case. What if the pods of the two jobs created/processed alternatively? For example, batch scheduler processes pod 1 of job 1 and then processes pod 1 of job 2, then pod 2 of job 1, then pod 2 of job 2, and so on.

k82cn replied:
in kube-batch, it'll do allocation in job level (FCFS also for job); so for this case, we will not handle pod2 of job2 until all pod in job1 are handled :)

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+   * bind all Pods to hosts until job is ready: if `numMembers` <= `allocated Pods` + `pending Pods`, it's ready when `numMembers` <= `allocated Pods`; otherwise, `numMember` <= `allocated Pods` + `succeeded Pods`
+4. If can not allocate enough resources to the job, the pods stay pending; and the resource cannot be allocated to other job
+
+That may make resources (less than job's resource request) idle for a while, e.g. a huge job. The solution, e.g. backfill other smaller jobs to improve the resource utilization, will be proposed in coming release. In `allocate` phase, only pod's `NodeAffinity` takes effect; the other predicates/priorities will be included on-demand in coming release.
```
backfilling soungs too optimistic. many k8s clusters may not have that many jobs to backfill.

k82cn replied:
it's only one of solutions :)

vishh commented on contributors/design-proposals/scheduling/coscheduling.md:
```diff
@@ -0,0 +1,180 @@
+
+### Admission Controller
+
+If quota runs out in the middle of creating a group of pods, a few members of a `PodGroup` may be created, while the rest will be denied by the `ResourceQuota` admission controller. `.spec.TotalResource` is added in `PodGroup` to address this problem. When a `PodGroup` is created with `.spec.TotalResource`, so much quota is reserved for the group if there is available quota. Pods of group use the ...
```
Do mention the changes required to the default quota controller to make this proposal work. That will be changes to the core project and needs to be acceptable to other quota subsystem owners.

justaugustus commented:
/kind kep

liggitt commented:
just to verify, there are no in-tree changes proposed now, correct? this was a question in sig-arch while going through the features targeted at 1.13

k82cn commented:
> just to verify, there are no in-tree changes proposed now, correct?

@liggitt , yes, we do **NOT** have code changes in-tree for this feature in 1.13.

bsalamat commented:
/assign

justaugustus commented:
REMINDER: KEPs are moving to [k/enhancements](https://github.com/kubernetes/enhancements) on November 30. Please attempt to merge this KEP before then to signal consensus.
For more details on this change, review [this thread](https://groups.google.com/d/msg/kubernetes-dev/LIkZoIqCT20/t515CV2mAgAJ).

Any questions regarding this move should be directed to that thread and not asked on GitHub.

justaugustus commented:
KEPs [have moved](https://github.com/kubernetes/enhancements/pull/638) to [k/enhancements](https://github.com/kubernetes/enhancements). 
This PR will be closed and any additional changes to this KEP should be submitted to k/enhancements.
For more details on this change, review [this thread](https://groups.google.com/d/msg/kubernetes-dev/LIkZoIqCT20/t515CV2mAgAJ).

Any questions regarding this move should be directed to that thread and not asked on GitHub.
/close

k82cn commented:
Moved to https://github.com/kubernetes/enhancements/pull/639
